### PR TITLE
fix: update grad_output quant to avoid redundant work

### LINKED ISSUE #1693: [MXFP8] grad_output is quantized columnwise even if weight doesn't require gradients.
https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html#Handling-transposes

Based on the diagram, 

Image

If weight is frozen (for eg. in LoRA setting), we can avoid quantizing `grad_output` in columnwise direction.

Using following patch seems to work
```patch
diff --git a/transformer_engine/pytorch/module/linear.py b/transformer_engine/pytorch/module/linear.py
index 83dc652c..3e58700a 100644
--- a/transformer_engine/pytorch/module/linear.py
+++ b/transformer_engine/pytorch/module/linear.py
@@ -423,6 +423,11 @@ class _Linear(torch.autograd.Function):
                     ub_obj_wgrad.set_buffer_params(ctx.grad_input_quantizer)
                     dgrad_bulk = ub_obj_wgrad.get_buffer(ctx.grad_input_quantizer)
 
+            if not ctx.requires_dgrad and ctx.grad_output_quantizer is not None:
+                ctx.grad_output_quantizer.set_usage(rowwise=False)
+            if not ctx.requires_wgrad and ctx.grad_output_quantizer is not None:
+                ctx.grad_output_quantizer.set_usage(columnwise=False)
+
             # Prepare grad output tensor
             # Note: Cast to expected dtype and perform tensor-parallel communication
             if ctx.grad_output_quantizer is not None:
```

Test Script
```python
import torch

import transformer_engine
from transformer_engine.pytorch import fp8_autocast, Linear

dim = 1024 * 22  # Large input for demonstration of memory change.
linear = Linear(dim, dim, bias=False)
x = torch.randn(dim, dim, requires_grad=True, device="cuda")

linear.weight.requires_grad = False

with fp8_autocast():
    o = linear(x)
    g_o = torch.randn_like(o)

o.backward(g_o)

# Without patch - 12314.476544 MB
# With patch - 11790.188544 MB
print(torch.cuda.max_memory_allocated() / 1e6, "MB")

```

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #1693

### CODE REVIEW

**Review by ksivaman — DISMISSED** [review]
LGTM

**Comment by IvanYashchuk** [comment]
Hi, @ksivaman, is there anything left to do to merge this PR?

File: tests/pytorch/test_sanity.py [hunk 1]
```diff
@@ -995,6 +995,31 @@ def backward(ctx, grad_output):
     torch.testing.assert_close(grad_checkpoint, grad_standard)
 
 
+def test_linear_frozen_weights_memory_default_recipe():
```
Comment (ksivaman) [inline tests/pytorch/test_sanity.py:999]:
```suggestion
@pytest.mark.skipif(not fp8_available, reason=reason_for_no_fp8)
def test_linear_frozen_weights_memory_default_recipe():
```
